Implements RateLimit for API Request/Response

simonneutert opened the pull request:
@dblock sorry for the delay, I am struggling a little at the time 😢 

I really hope this attempt is way more appealing. If so, I will try to finish it asap.  

This PR superseeds https://github.com/dblock/strava-ruby-client/pull/40 

closes #11

dblock commented on lib/strava/api/ratelimit.rb:
```diff
@@ -0,0 +1,92 @@
+        @headers = headers
+      end
+
+      def ratelimit_hash
```
Make it private and/or implement a public `to_h`?

dblock commented on lib/strava/api/ratelimit.rb:
```diff
@@ -0,0 +1,92 @@
+      def ratelimit_hash
+        if ratelimit?
+          {
+            'ratelimit_limit' => limit,
```
Keys should be symbols.

dblock commented on lib/strava/api/ratelimit.rb:
```diff
@@ -0,0 +1,92 @@
+      # @return [TrueClass]
+      #
+      def ratelimit?
+        !@headers['x-ratelimit-limit'].nil?
```
@headers.key?('x-ratelimit-limit')

dblock commented on lib/strava/api/ratelimit.rb:
```diff
@@ -0,0 +1,92 @@
+      end
+
+      def fiveteen_minutes
+        ratelimit? ? limit.split(',').first.to_i : 0
```
I think these should return `nil` when ratelimit not set, otherwise it might imply that the value parsed is 0. Make sure `limit` returns `nil`, then `limit?.split(',')...`

dblock commented on lib/strava/api/ratelimit.rb:
```diff
@@ -0,0 +1,92 @@
+      end
+
+      def fiveteen_minutes_remaining
+        fiveteen_minutes.to_i - fiveteen_minutes_usage.to_i
```
return `nil` if no limit everwhere

dblock commented on lib/strava/web/response.rb:
```diff
@@ -0,0 +1,31 @@
+        case response.body.class.name
+        when 'Array'
+          response.body.map! do |elem|
+            merge_ratelimit!(elem, response)
```
It's possible that `response.body` is frozen, so you cannot modify it in-place. Assign ratelimit instance to the response object.

dblock commented on spec/strava/api/client/endpoints/activities/activity_spec.rb:
```diff
@@ -5,6 +5,15 @@
+    expect(activity.ratelimit_total_day).to eq 30_000
+    expect(activity.ratelimit_total_day_remaining).to eq 22_776
+    expect(activity.ratelimit_total_day_usage).to eq 7224
+    expect(activity.ratelimit_usage).to eq '1,7224'
```
I think I want to write `activity.http_response.ratelimit`.

dblock reviewed (changes requested):
Thanks! 

I quite dislike that we're extending response classes with the ratelimit information. It's not future-proof for other information returned in headers. I'd like to be able to do `activity._http_response.rate_limit` (and `activity._http_request...` similarly).

simonneutert commented:
@dblock I gave it another go 🤞

simonneutert commented:
tests run clear ✅

dblock commented:
Let's get https://github.com/dblock/strava-ruby-client/pull/52 first in!

dblock commented:
Care to rebase this @simonneutert?

simonneutert commented:
it must feel like kindergarten for you, but be assured I appreciate the time you take to teach me so very much! 
pulled master back in 👍

dblock commented on .github/workflows/ruby.yml:
```diff
@@ -0,0 +1,31 @@
+        ruby-version: ${{ matrix.ruby-version }}
+        bundler-cache: true # runs 'bundle install' and caches installed gems automatically
+    - name: Run tests
+      run: RACK_ENV=test bundle exec rake
```
This file was renamed earlier to `test.yml`, so `git rm` it.

dblock commented on lib/strava/web/request.rb:
```diff
@@ -35,7 +35,23 @@ def request(method, path, options)
+      end
+
+      def conditional_response_upgrade(response)
+        case response.body.class.name
```
Does it work with just `.class` and `case Array`? Avoids conversion to string. I would just use `if` and `is_a?`.

dblock commented on lib/strava/web/request.rb:
```diff
@@ -35,7 +35,23 @@ def request(method, path, options)
+      def conditional_response_upgrade(response)
+        case response.body.class.name
+        when 'Array'
+          response.dup.body.map! do |body_elem|
```
`response.body.dup`, avoid duplicating the entire response

dblock commented on lib/strava/web/request.rb:
```diff
@@ -35,7 +35,23 @@ def request(method, path, options)
+            body_elem
+          end
+        when 'Hash'
+          response.dup.body['http_response'] = response
```
same as above 2x

dblock commented on lib/strava/web/response_wrap.rb:
```diff
@@ -0,0 +1,40 @@
+        if @response.instance_of?(Array)
+          @response.send(method_symbol, *args, &block)
+        else
+          case @response.body.class.name
```
`body.class`, or use `is_a?`

dblock commented on lib/strava/web/response_wrap.rb:
```diff
@@ -0,0 +1,40 @@
+
+module Strava
+  module Web
+    class ResponseWrap
```
Can we just call it `Response`?

dblock commented on lib/strava/webhooks/models/subscription.rb:
```diff
@@ -4,6 +4,8 @@ module Strava
   module Webhooks
     module Models
       class Subscription < Hashie::Trash
+        include Strava::Models::Mixins::HttpResponse
```
Can we have this automatically added to any response? Make a base class that includes `HttpResponse` that inherits from `Hashie::Trash` and make all models inherit that.

dblock reviewed (changes requested):
Some more comments. Add to CHANGELOG please and documentation in README.

adamwolf commented:
Hi folks! I'm wondering where this has landed.  If this is something that I can help finish up, I'd love to try to help!

simonneutert commented:
Hi @adamwolf what needs to be done was commented. I'd love to have someone work with me on that one 👍

adamwolf commented:
I've addressed much of the PR feedback at https://github.com/adamwolf/strava-ruby-client/tree/rate-limit. I added commits after simonneutert's work.

I'm not 100% sure of my changes (and I haven't made any changes to README.md), but the commit messages should explain which things I'm not really sure of.  Tests pass!

I'm pretty handy with git, so feel free to ask me to rebase or squash or do whatever.  I do not intend to take any credit away from simonneutert or amyone--I'm just a person hitting rate limit with a Strava integration :)

simonneutert commented:
If you finish it, I am glad to have helped paving the start. The credit is all yours then 🤝

So happy you jumped in to help 🥳

dblock commented:
@adamwolf please open a PR!

simonneutert commented:
I've took the freedom and pulled the PR of @adamwolf and rebased to current master 🤞 

As far as I can tell there is no model, that doesn't inherit from the new `Strava::Web::Reponse` class.

- [x] added a spec which can be rolled out to the rest of the models, if it's pleasing.
- [x] added a stub for the Readme.md
- [x] added a stub for Upgrade.md

I couldn't figure out what you mean [here](https://github.com/dblock/strava-ruby-client/pull/61#pullrequestreview-1017767754) @dblock so please give me a clue.

adamwolf commented:
Thanks! I had some other things come up at work and they ended up bringing us under the rate limit, so this got preempted a bit... :)

dblock commented on .github/workflows/ruby.yml:
```diff
@@ -0,0 +1,31 @@
+# This workflow uses actions that are not certified by GitHub.
```
This file exists as https://github.com/dblock/strava-ruby-client/blob/master/.github/workflows/test.yml, delete this dup.

dblock reviewed (changes requested):
I've unresolved some conversations/questions I had.
